File: backend/app/schemas/referral.py
"""Referral Management and Tracking Pydantic Schemas."""
from datetime import datetime
from enum import Enum
from typing import Optional, List, Dict, Any
from uuid import UUID
from pydantic import BaseModel, Field, ConfigDict


class ReferralStatus(str, Enum):
    """Permitted states in the referral lifecycle state machine."""
    CREATED = "CREATED"
    PENDING_ACCEPTANCE = "PENDING_ACCEPTANCE"
    ACCEPTED = "ACCEPTED"
    REJECTED = "REJECTED"
    PATIENT_NOTIFIED = "PATIENT_NOTIFIED"
    DEPARTED = "DEPARTED"
    ARRIVED = "ARRIVED"
    IN_SERVICE = "IN_SERVICE"
    COMPLETED = "COMPLETED"
    BACK_REFERRED = "BACK_REFERRED"
    FOLLOW_UP = "FOLLOW_UP"
    CLOSED = "CLOSED"
    REROUTED = "REROUTED"
    CANCELLED = "CANCELLED"


class RejectionReason(str, Enum):
    """Controlled reasons for receiving facility referral rejection."""
    SERVICE_UNAVAILABLE = "SERVICE_UNAVAILABLE"
    CAPACITY_UNAVAILABLE = "CAPACITY_UNAVAILABLE"
    SPECIALIST_UNAVAILABLE = "SPECIALIST_UNAVAILABLE"
    FACILITY_CLOSED = "FACILITY_CLOSED"
    OTHER = "OTHER"


class ReferralUrgency(str, Enum):
    """Referral clinical urgency level."""
    EMERGENCY = "EMERGENCY"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


# -------------------------------------------------------------------------
# Request Schemas
# -------------------------------------------------------------------------

class ReferralCreate(BaseModel):
    """Schema for initiating a new referral from a Care Request and Facility."""
    care_request_id: UUID = Field(..., description="Target Care Request UUID")
    receiving_facility_id: UUID = Field(..., description="Target Receiving Facility UUID")
    source_facility_id: Optional[UUID] = Field(None, description="Optional Origin Facility UUID")
    referral_reason: Optional[str] = Field(None, max_length=2000, description="Clinical reason for referral")
    clinical_summary: Optional[str] = Field(None, max_length=4000, description="Detailed clinical summary / provisional diagnosis")
    urgency: Optional[str] = Field("MEDIUM", description="Referral urgency level (EMERGENCY, HIGH, MEDIUM, LOW)")
    required_specialty: Optional[str] = Field(None, max_length=100, description="Required clinical specialty")
    required_capability: Optional[str] = Field(None, max_length=100, description="Specific service capability required")
    expected_arrival_time: Optional[datetime] = Field(None, description="Expected patient arrival timestamp")
    transport_mode: Optional[str] = Field(None, description="Transport mode (108_AMBULANCE, 102_AMBULANCE, PRIVATE_VEHICLE, etc.)")


class ReferralAcceptInput(BaseModel):
    """Schema for facility acceptance action."""
    notes: Optional[str] = Field(None, max_length=1000, description="Acceptance notes or admission instructions")
    expected_arrival_time: Optional[datetime] = Field(None, description="Updated expected arrival timestamp")


class ReferralRejectInput(BaseModel):
    """Schema for facility rejection action."""
    rejection_reason: str = Field(..., description="Controlled rejection reason (SERVICE_UNAVAILABLE, CAPACITY_UNAVAILABLE, etc.)")
    rejection_notes: Optional[str] = Field(None, max_length=1000, description="Detailed reason notes")


class ReferralNotifyInput(BaseModel):
    """Schema for patient notification action."""
    notes: Optional[str] = Field(None, max_length=1000, description="Notification method or counseling notes")


class ReferralDepartInput(BaseModel):
    """Schema for patient departure action."""
    transport_mode: Optional[str] = Field(None, max_length=100, description="Transport vehicle / ambulance ID or mode")
    estimated_transit_minutes: Optional[int] = Field(None, ge=1, le=1440, description="Estimated transit time in minutes")
    notes: Optional[str] = Field(None, max_length=1000, description="Departure remarks")


class ReferralArriveInput(BaseModel):
    """Schema for patient arrival action."""
    notes: Optional[str] = Field(None, max_length=1000, description="Triage or reception remarks upon arrival")


class ReferralStartServiceInput(BaseModel):
    """Schema for consultation or service initiation action."""
    notes: Optional[str] = Field(None, max_length=1000, description="Treating doctor or department notes")


class ReferralCompleteInput(BaseModel):
    """Schema for service completion action."""
    clinical_summary: Optional[str] = Field(None, max_length=4000, description="Treatment summary, discharge note, or outcome")
    notes: Optional[str] = Field(None, max_length=1000, description="General completion notes")


class ReferralBackReferInput(BaseModel):
    """Schema for issuing back-referral instructions to primary health center."""
    back_referral_notes: str = Field(..., min_length=5, max_length=4000, description="Clinical guidance and follow-up instructions for primary care worker")
    notes: Optional[str] = Field(None, max_length=1000, description="Additional administrative remarks")


class ReferralRerouteInput(BaseModel):
    """Schema for rerouting a rejected referral to an alternative facility."""
    new_receiving_facility_id: UUID = Field(..., description="Alternative receiving facility UUID")
    reason: Optional[str] = Field(None, max_length=1000, description="Rerouting justification")
    notes: Optional[str] = Field(None, max_length=1000, description="Additional clinical notes for new facility")


# -------------------------------------------------------------------------
# Response Schemas
# -------------------------------------------------------------------------

class ReferralEventResponse(BaseModel):
    """Immutable referral lifecycle event log entry."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    referral_id: UUID
    event_type: str
    previous_status: Optional[str] = None
    new_status: str
    performed_by: Optional[UUID] = None
    performer_name: Optional[str] = None
    performer_role: Optional[str] = None
    notes: Optional[str] = None
    event_metadata: Optional[Dict[str, Any]] = None
    created_at: datetime


class ReferralResponse(BaseModel):
    """Detailed referral record response schema."""
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    referral_code: str
    care_request_id: UUID
    care_request_number: Optional[str] = None
    care_category: Optional[str] = None
    required_service: Optional[str] = None

    patient_id: UUID
    patient_name: Optional[str] = None
    patient_code: Optional[str] = None
    patient_age: Optional[int] = None
    patient_gender: Optional[str] = None
    patient_phone: Optional[str] = None
    patient_village: Optional[str] = None

    source_facility_id: Optional[UUID] = None
    source_facility_name: Optional[str] = None
    receiving_facility_id: UUID
    receiving_facility_name: Optional[str] = None
    receiving_facility_type: Optional[str] = None

    created_by: Optional[UUID] = None
    creator_name: Optional[str] = None
    creator_role: Optional[str] = None
    parent_referral_id: Optional[UUID] = None

    status: str
    urgency: str
    referral_reason: Optional[str] = None
    clinical_summary: Optional[str] = None
    required_specialty: Optional[str] = None

    transport_mode: Optional[str] = None
    transport_status: Optional[str] = None
    estimated_transit_minutes: Optional[int] = None
    expected_arrival_time: Optional[datetime] = None

    rejection_reason: Optional[str] = None
    rejection_notes: Optional[str] = None
    back_referral_notes: Optional[str] = None

    initiated_at: datetime
    accepted_at: Optional[datetime] = None
    notified_at: Optional[datetime] = None
    departed_at: Optional[datetime] = None
    arrived_at: Optional[datetime] = None
    in_service_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    back_referred_at: Optional[datetime] = None
    closed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    events: Optional[List[ReferralEventResponse]] = None


class ReferralListResponse(BaseModel):
    """Paginated list of referrals."""
    items: List[ReferralResponse]
    total: int
    page: int
    page_size: int
    total_pages: int

File: frontend/src/types/referral.ts
/**
 * Referral Lifecycle and Tracking TypeScript Definitions.
 */

export type ReferralStatus =
  | "CREATED"
  | "PENDING_ACCEPTANCE"
  | "ACCEPTED"
  | "REJECTED"
  | "PATIENT_NOTIFIED"
  | "DEPARTED"
  | "ARRIVED"
  | "IN_SERVICE"
  | "COMPLETED"
  | "BACK_REFERRED"
  | "FOLLOW_UP"
  | "CLOSED"
  | "REROUTED"
  | "CANCELLED";

export type RejectionReason =
  | "SERVICE_UNAVAILABLE"
  | "CAPACITY_UNAVAILABLE"
  | "SPECIALIST_UNAVAILABLE"
  | "FACILITY_CLOSED"
  | "OTHER";

export type ReferralUrgency = "EMERGENCY" | "HIGH" | "MEDIUM" | "LOW";

export interface ReferralEvent {
  id: string;
  referral_id: string;
  event_type: string;
  previous_status?: string | null;
  new_status: string;
  performed_by?: string | null;
  performer_name?: string | null;
  performer_role?: string | null;
  notes?: string | null;
  event_metadata?: Record<string, unknown> | null;
  created_at: string;
}

export interface Referral {
  id: string;
  referral_code: string;
  care_request_id: string;
  care_request_number?: string | null;
  care_category?: string | null;
  required_service?: string | null;

  patient_id: string;
  patient_name?: string | null;
  patient_code?: string | null;
  patient_age?: number | null;
  patient_gender?: string | null;
  patient_phone?: string | null;
  patient_village?: string | null;

  source_facility_id?: string | null;
  source_facility_name?: string | null;
  receiving_facility_id: string;
  receiving_facility_name?: string | null;
  receiving_facility_type?: string | null;

  created_by?: string | null;
  creator_name?: string | null;
  creator_role?: string | null;
  parent_referral_id?: string | null;

  status: ReferralStatus;
  urgency: string;
  referral_reason?: string | null;
  clinical_summary?: string | null;
  required_specialty?: string | null;

  transport_mode?: string | null;
  transport_status?: string | null;
  estimated_transit_minutes?: number | null;
  expected_arrival_time?: string | null;

  rejection_reason?: string | null;
  rejection_notes?: string | null;
  back_referral_notes?: string | null;

  initiated_at: string;
  accepted_at?: string | null;
  notified_at?: string | null;
  departed_at?: string | null;
  arrived_at?: string | null;
  in_service_at?: string | null;
  completed_at?: string | null;
  back_referred_at?: string | null;
  closed_at?: string | null;
  created_at: string;
  updated_at: string;

  events?: ReferralEvent[] | null;
}

export interface ReferralListResponse {
  items: Referral[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ReferralCreateInput {
  care_request_id: string;
  receiving_facility_id: string;
  source_facility_id?: string | null;
  referral_reason?: string | null;
  clinical_summary?: string | null;
  urgency?: string;
  required_specialty?: string | null;
  required_capability?: string | null;
  expected_arrival_time?: string | null;
  transport_mode?: string | null;
}

export interface ReferralAcceptInput {
  notes?: string;
  expected_arrival_time?: string;
}

export interface ReferralRejectInput {
  rejection_reason: string;
  rejection_notes?: string;
}

export interface ReferralNotifyInput {
  notes?: string;
}

export interface ReferralDepartInput {
  transport_mode?: string;
  estimated_transit_minutes?: number;
  notes?: string;
}

export interface ReferralArriveInput {
  notes?: string;
}

export interface ReferralStartServiceInput {
  notes?: string;
}

export interface ReferralCompleteInput {
  clinical_summary?: string;
  notes?: string;
}

export interface ReferralBackReferInput {
  back_referral_notes: string;
  notes?: string;
}

export interface ReferralRerouteInput {
  new_receiving_facility_id: string;
  reason?: string;
  notes?: string;
}
